quickwit/quickwit-storage/src/local_file_storage.rs

```diff
@@ -180,14 +180,24 @@ impl Storage for LocalFileStorage {
         payload: Box<dyn crate::PutPayload>,
     ) -> crate::StorageResult<()> {
         let full_path = self.full_path(path)?;
-        if let Some(parent_dir) = full_path.parent() {
-            fs::create_dir_all(parent_dir).await?;
-        }
+        let parent_dir = full_path.parent()
+            .ok_or_else(|| {
+                let err = anyhow::anyhow!("No parent directory for {full_path:?}");
+                StorageErrorKind::InternalError.with_error(err)
+            })?;
 
+        fs::create_dir_all(parent_dir).await?;
         let mut reader = payload.byte_stream().await?.into_async_read();
-        let mut f = tokio::fs::File::create(full_path).await?;
-        tokio::io::copy(&mut reader, &mut f).await?;
-
+        let named_temp_file = tempfile::NamedTempFile::new_in(parent_dir)?;
+        let (temp_std_file, temp_filepath) = named_temp_file.into_parts();
+        let mut temp_tokio_file = tokio::fs::File::from_std(temp_std_file);
+        tokio::io::copy(&mut reader, &mut temp_tokio_file).await?;
+        temp_tokio_file.flush().await?;
+        temp_tokio_file.sync_data().await?;
+        temp_filepath
+            .persist(&full_path)
+            .map_err(|err| StorageErrorKind::Io.with_error(err))?;
+        tokio::fs::File::open(parent_dir).await?.sync_data().await?;
```

Comment:
Nitpick: For readability purposes, I think we should use `OpenOptions` rather than `File::open()` to avoid confusion, even though I understand why we're opening this directory, it feels odd opening this a `File`.